draft-ietf-quic-recovery.md

```diff
@@ -1214,15 +1219,27 @@ Invoked by loss detection from DetectLostPackets when new packets
 are detected lost.
 
 ~~~
+   InPersistentCongestion(oldest_loss_time):
+     congestion_period = Now() - oldest_loss_time
+     pto = smoothed_rtt + 4 * rttvar + max_ack_delay
+     return
+       congestion_period >
+         pto * (2 ^ kPersistentCongestionThreshold - 1)
+
    OnPacketsLost(lost_packets):
      // Remove lost packets from bytes_in_flight.
      for (lost_packet : lost_packets):
        bytes_in_flight -= lost_packet.size
-     largest_lost_packet = lost_packets.last()
+     oldest_lost_packet = lost_packets.first()
+     newest_lost_packet = lost_packets.last()
 
      // Start a new congestion epoch if the last lost packet
      // is past the end of the previous recovery epoch.
-     CongestionEvent(largest_lost_packet.time_sent)
+     CongestionEvent(newest_lost_packet.time_sent)
+
+     // Collapse congestion window if persistent congestion
+     if (InPersistentCongestion(oldest_lost_packet.time_sent)):
```

Comment:
I'm starting to think we should pass in congestion_period and make it oldest_lost_packet - newest_lost_packet.

I think using Now() may be incorrect.  What if one packet is lost and then a few ACKs are lost.  In this case, Now() is substantially larger than oldest_lost_packet, but in fact only one packet was lost.

Reply:
I'd like the opinion of @janaiyengar here.  I think this is likely fairly good, but I think passing in newest_lost_time is easy and more correct, given they're easily available in the pseudocode.

Reply:
Let me propose an example. Assume smooth_rtted = 1 sec, rttvar = 0, max_ack_delay = 0 for simplicity's sake.
```
t=0 | Send Pkt #1 (App Data)
t=1 | Send Pkt #2 (PTO 1)
t=3 | Send Pkt #3 (PTO 2)
t=7 | Send Pkt #4 (PTO 3)
t=8 | Recv ACK of Pkt #4
```
At t=8 you'd get a congestion event for packets 1 & 2 being lost. If you use Now() for the persistent congestion calculation, this would be declared as persistent congestion. If you used newest_lost_time it wouldn't. Which behavior do we want?

Reply:
If 4 is acked and the others are lost, then it should be persistent congestion.

But if 2, 3 and 4 were all acked at once(t=8), that should not be persistent congestion.

Reply:
Hmm, well then we'd have to make two changes to achieve that behavior.
1. Use `newest_lost_packet - oldest_lost_packet` for the time window.
2. Make the threshold 2 PTO instead of 3.

Without the second part above, an ACK of pkt 4 would produce a loss window of only 3 seconds (3 - 0), right?

Reply:
That sounds correct to me.  Can you add your example as an example to the text section as well?  This is clearly very confusing, so an example would help.

Reply:
Sure. Do you want me to describe the example in just words, or put almost exactly what I have above with the text?

Reply:
I think how you have it laid out is great and an effort to convert it into text will probably make it less clear.  But I'm not familiar enough with markdown to know how to get the formatting right.

Reply:
Ok. I'll give it a go.